### DISCUSSION TITLE
Single file for data content collections

### DISCUSSION THREAD

**Discussion by michaelmcleodnz** [withastro/roadmap#626] [Proposal]
# Summary

The ability add a JSON file to `src/content` that defines an entire collection, instead of requiring a folder containing many separate JSON files.

# Background & Motivation

I would like to create a wiki-style site, where each individual page has a relatively small amount of data, but the number of pages is large. I imagine this would also be the case for a shopping site that wanted many small pages for each product, but had a potentially large number of products.

My data doesn't change very often, and so it wouldn't need anything to happen server-side. But writing many similar HTML pages by hand would be awful, so a tool like Astro that can generate the HTML from a data source would be perfect. In fact, it can already do this by importing a JSON file and using it with `getStaticPaths`. It seems like the most sensible place for this data would be the `src/content` directory, where the schema could be defined using `config.ts`, but right now that would require splitting the data into many small JSON files, one for each entry, and editing the entries would require navigating between a large number of small files instead of just editing all the data in one file.

# Goals

- Have all content for a collection defined in a single file, for the cases where this makes sense.
- This would be a good first step towards implementing other suggestions like [dynamic collections](https://github.com/withastro/roadmap/discussions/452), that would necessarily have all collection information stored in a single file and not a directory.

# Example

I imagine it would involve a change to the `defineCollection` function, so that you could indicate that the data was in a file and not a directory, and that collections of this type would require you to manually define the field used for slugs and references.

__config.ts__
```diff
import { z, defineCollection } from 'astro:content';

const productsCollection = defineCollection({
+ type: 'file',            // or 'collection' or 'multidata' or 'list' ?
  schema: z.object({
    name: z.string(),
    price: z.number(),
    // ...
  }),
+ referenceField: 'id',    // or slug or id or idField ?
});

export const collections = {
  'products': productsCollection,
};
```

__src/content/products.json__
```javascript
[
  {
    "id": 1,
    "name": "Funny hat",
    "price": 1.00,
    // ...
  },
  // ...
]
```

**Comment by florian-lefebvre** [comment]
That's possible since 5.0: https://docs.astro.build/en/reference/content-loader-reference/#file-loader
